File: tests/test_arithmetic.mjs
// Conformance test — JS implementation of Δ/Δd algebra (spec §11.4)
// against docs/spec/test-vectors-arithmetic.json.
//
// Run with:  node tests/test_arithmetic.mjs
//
// Loads docs/assets/js/ats.js into a `node:vm` context (the IIFE binds
// to `globalThis.ATS` when `window` is absent), avoiding duplication.

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';
import vm from 'node:vm';
import assert from 'node:assert/strict';

const here = dirname(fileURLToPath(import.meta.url));
const repo = resolve(here, '..');

const sourceCode = readFileSync(resolve(repo, 'docs/assets/js/ats.js'), 'utf-8');
const ctx = { console };
vm.createContext(ctx);
vm.runInContext(sourceCode, ctx);
const ATS = ctx.ATS;
assert.ok(ATS && typeof ATS.add === 'function', 'ATS.add must be exposed from ats.js');

const ATS_SCALE = 100000;

function parseAts(str) {
  const m = /^\s*(T[+-])\s*Δ\s+(\d+)\.(\d)\.(\d)\.(\d)\.(\d+)\s*$/.exec(str);
  if (!m) throw new Error(`Bad canonical: ${str}`);
  const fracStr = m[6].slice(0, 5).padEnd(5, '0');
  return {
    sign: m[1],
    kilo: parseInt(m[2], 10),
    hecto: parseInt(m[3], 10),
    deka: parseInt(m[4], 10),
    kin: parseInt(m[5], 10),
    frac: parseInt(fracStr, 10),
  };
}

function materialize(op) {
  if (op.kind === 'ats') return parseAts(op.canonical);
  if (op.kind === 'dur') {
    if (op.signed_days !== undefined) return ATS.dur(Number(op.signed_days));
    return ATS.durFromCanonical(op.canonical);
  }
  if (op.kind === 'scalar') return op.value;
  if (op.kind === 'bool') return op.value;
  throw new Error(`Unknown kind: ${op.kind}`);
}

function runOp(op, lhs, rhs) {
  switch (op) {
    case 'add': return ATS.add(lhs, rhs);
    case 'sub': return ATS.sub(lhs, rhs);
    case 'mul': return ATS.mul(lhs, rhs);
    case 'div': return ATS.div(lhs, rhs);
    case 'lt':  return ATS.lt(lhs, rhs);
    case 'le':  return ATS.le(lhs, rhs);
    case 'eq':  return ATS.eq(lhs, rhs);
    case 'gt':  return ATS.gt(lhs, rhs);
    case 'ge':  return ATS.ge(lhs, rhs);
    default: throw new Error(`Unknown op: ${op}`);
  }
}

function expectedCanonical(expected) {
  if (expected.kind === 'ats') return expected.canonical;
  if (expected.kind === 'dur') {
    if (expected.canonical) return expected.canonical;
    return ATS.durToCanonical(ATS.dur(Number(expected.signed_days)));
  }
  throw new Error(`No canonical for ${expected.kind}`);
}

const vectors = JSON.parse(
  readFileSync(resolve(repo, 'docs/spec/test-vectors-arithmetic.json'), 'utf-8'),
).vectors;

let pass = 0, fail = 0;
for (const v of vectors) {
  const lhs = materialize(v.lhs);
  const rhs = materialize(v.rhs);
  try {
    const got = runOp(v.op, lhs, rhs);
    if (v.expected.kind === 'bool') {
      assert.equal(got, v.expected.value, v.label);
    } else {
      const gotCanonical = v.expected.kind === 'ats'
        ? ATS.toCanonical(got)
        : ATS.durToCanonical(got);
      assert.equal(gotCanonical, expectedCanonical(v.expected), v.label);
    }
    pass += 1;
  } catch (e) {
    fail += 1;
    console.error('FAIL', v.label, '\n ', e.message);
  }
}

// Extra: sign-zero equality + duration neg/abs round trip.
const plusZero = parseAts('T+ Δ 0.0.0.0.00000');
const minusZero = parseAts('T- Δ 0.0.0.0.00000');
try {
  assert.equal(ATS.eq(plusZero, minusZero), true, 'T+0 == T-0');
  const d = ATS.dur(1.5);
  assert.equal(ATS.neg(d).signedDays, -1.5, '-Δd');
  assert.equal(ATS.abs(ATS.neg(d)).signedDays, 1.5, '|−Δd|');
  pass += 3;
} catch (e) {
  fail += 1;
  console.error('FAIL extras\n ', e.message);
}

console.log(`${pass}/${vectors.length + 3} JS arithmetic vectors pass`);
if (fail) process.exit(1);

File: tests/test_arithmetic.py
"""Conformance test — Δ/Δd algebra (spec §11.4, v0.6+).

Drives the same JSON-defined vectors as the JS conformance suite. The
JSON format uses `{kind, canonical|signed_days|value}` shapes:

  ats:    {"kind": "ats",    "canonical": "T+ Δ 20.7.8.2.50000"}
  dur:    {"kind": "dur",    "signed_days": "1.5"}            (input)
  dur:    {"kind": "dur",    "canonical": "T+ Δd 0.0.0.1.50000"} (expected)
  scalar: {"kind": "scalar", "value": 2}
  bool:   {"kind": "bool",   "value": true}

Operations: add, sub, mul, div, lt, le, eq, gt, ge.

Run with:  python -m unittest tests.test_arithmetic
"""

from __future__ import annotations

import json
import re
import sys
import unittest
from decimal import Decimal
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT / "code"))

from ats import (  # noqa: E402
    ATS_DECIMALS,
    ATSDateTime,
    ATSDuration,
)

VECTORS_PATH = ROOT / "docs" / "spec" / "test-vectors-arithmetic.json"

_CANON_RE = re.compile(r"^\s*(T[+-])\s*Δ\s+(\d+)\.(\d)\.(\d)\.(\d)\.(\d+)\s*$")
_DUR_RE = re.compile(r"^\s*(T[+-])\s*Δd\s+(\d+)\.(\d)\.(\d)\.(\d)\.(\d+)\s*$")


def parse_canonical(s: str) -> ATSDateTime:
    m = _CANON_RE.match(s)
    if not m:
        raise ValueError(f"Bad canonical: {s!r}")
    sign, kilo, hecto, deka, kin, frac = m.groups()
    return ATSDateTime(
        sign=sign,
        kilo=int(kilo),
        hecto=int(hecto),
        deka=int(deka),
        kin=int(kin),
        frac=int(frac.ljust(ATS_DECIMALS, "0")[:ATS_DECIMALS]),
    )


def parse_duration_canonical(s: str) -> ATSDuration:
    m = _DUR_RE.match(s)
    if not m:
        raise ValueError(f"Bad Δd canonical: {s!r}")
    sign, kilo, hecto, deka, kin, frac = m.groups()
    integer_days = int(kilo) * 1000 + int(hecto) * 100 + int(deka) * 10 + int(kin)
    frac_padded = frac.ljust(ATS_DECIMALS, "0")[:ATS_DECIMALS]
    abs_days = Decimal(integer_days) + Decimal(int(frac_padded)) / Decimal(10**ATS_DECIMALS)
    return ATSDuration(-abs_days if sign == "T-" else abs_days)


def materialize(operand: dict):
    kind = operand["kind"]
    if kind == "ats":
        return parse_canonical(operand["canonical"])
    if kind == "dur":
        if "signed_days" in operand:
            return ATSDuration(Decimal(operand["signed_days"]))
        return parse_duration_canonical(operand["canonical"])
    if kind == "scalar":
        return operand["value"]
    if kind == "bool":
        return operand["value"]
    raise ValueError(f"Unknown operand kind: {kind}")


def expected_canonical(operand: dict) -> str:
    if operand["kind"] == "ats":
        return operand["canonical"]
    if operand["kind"] == "dur":
        if "canonical" in operand:
            return operand["canonical"]
        # signed_days expected — build the canonical via ATSDuration.
        return ATSDuration(Decimal(operand["signed_days"])).to_canonical()
    raise ValueError(f"No canonical for kind {operand['kind']!r}")


class TestArithmetic(unittest.TestCase):
    def setUp(self) -> None:
        with VECTORS_PATH.open(encoding="utf-8") as fh:
            self.vectors = json.load(fh)["vectors"]

    def _run_op(self, op: str, lhs, rhs):
        if op == "add":
            return lhs + rhs
        if op == "sub":
            return lhs - rhs
        if op == "mul":
            return lhs * rhs
        if op == "div":
            return lhs / rhs
        if op == "lt":
            return lhs < rhs
        if op == "le":
            return lhs <= rhs
        if op == "eq":
            return lhs == rhs
        if op == "gt":
            return lhs > rhs
        if op == "ge":
            return lhs >= rhs
        raise ValueError(f"Unknown op: {op}")

    def test_each_vector(self) -> None:
        for v in self.vectors:
            with self.subTest(label=v["label"]):
                lhs = materialize(v["lhs"])
                rhs = materialize(v["rhs"])
                expected = v["expected"]
                got = self._run_op(v["op"], lhs, rhs)
                if expected["kind"] == "bool":
                    self.assertEqual(got, expected["value"])
                else:
                    self.assertEqual(got.to_canonical(), expected_canonical(expected))

    def test_duration_neg_and_abs(self) -> None:
        d = ATSDuration(Decimal("1.5"))
        self.assertEqual((-d).signed_days, Decimal("-1.5"))
        self.assertEqual(abs(-d).signed_days, Decimal("1.5"))

    def test_comparison_equality_across_zero_signs(self) -> None:
        plus_zero = ATSDateTime(sign="T+", kilo=0, hecto=0, deka=0, kin=0, frac=0)
        minus_zero = ATSDateTime(sign="T-", kilo=0, hecto=0, deka=0, kin=0, frac=0)
        self.assertEqual(plus_zero, minus_zero)
        self.assertFalse(plus_zero < minus_zero)
        self.assertFalse(plus_zero > minus_zero)

    def test_duration_round_trip_within_resolution(self) -> None:
        d = ATSDuration(Decimal("123.45678"))
        canonical = d.to_canonical()
        # Re-parse and verify floor preserves up to ATS_DECIMALS digits.
        parsed = parse_duration_canonical(canonical)
        diff = abs(d.signed_days - parsed.signed_days)
        self.assertLess(diff, Decimal(1) / Decimal(10**ATS_DECIMALS))


if __name__ == "__main__":
    unittest.main()
